File: Server/Tuckshop.IdentityServer/Tuckshop.IdentityServer.Contracts/Models/UserManagement/Queries/UserLookup.cs
namespace Tuckshop.IdentityServer.Contracts.UserManagement.Queries
{
  using System;

  /// <summary>
  /// User lookup.
  /// </summary>
  public class UserLookup
  {
    /// <summary>
    /// Gets or sets the user id.
    /// </summary>
    public string Id { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the First Name starts.
    /// </summary>
    public string FirstName { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the Last Name starts.
    /// </summary>
    public string LastName { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the User Name.
    /// </summary>
    public string UserName { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets a value indicating whether the user Is Active.
    /// </summary>
    public bool IsActive { get; set; }

    /// <summary>
    /// Gets or sets a value indicating whether the user has a Email is Confirmed.
    /// </summary>
    public bool EmailConfirmed { get; set; }

    /// <summary>
    /// Gets or sets a value indicating whether the user Two Factor is Enabled.
    /// </summary>
    public bool TwoFactorEnabled { get; set; }

    /// <summary>
    /// Gets or sets a value indicating whether the user Two Factor is Configured.
    /// </summary>
    public bool TwoFactorConfigured { get; set; }

    /// <summary>
    /// Gets or sets the users lockout end date.
    /// </summary>
    public DateTime? LockoutEnd { get; set; }

    /// <summary>
    /// Gets or sets the Identity Provider name.
    /// </summary>
    public string IdentityProvider { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets a value indicating whether the identity provider is external.
    /// </summary>
    public bool IsExternalIdentityProvider { get; set; }

    /// <summary>
    /// Gets or sets a value indicating whether the tenant identity provider requires MFA.
    /// </summary>
    public bool ProviderRequiresMFA { get; set; }
  }
}

File: Client/Tuckshop.WebApp/src/Identity/Models/UserManagement/Queries/UserLookup.ts
import { Attributes, ModelBase, NeoModel, StringUtils } from '@singularsystems/neo-core';

@NeoModel
export default class UserLookup extends ModelBase {

    public id: string = "";

    public firstName: string = "";

    public lastName: string = "";

    public userName: string = "";

    public isActive: boolean = false;

    public emailConfirmed: boolean = false;

    public twoFactorEnabled: boolean = false;

    public twoFactorConfigured: boolean = false;

    @Attributes.Date()
    public lockoutEnd: Date | null = null;

    public identityProvider: string = "";

    public isExternalIdentityProvider: boolean = false;

    public providerRequiresMFA: boolean = false;

    // Client only properties / methods

    public get canSendVerificationLink() {
        return !this.isExternalIdentityProvider && !this.emailConfirmed && this.isActive;
    }

    public get canEnableMFA() {
        return !this.isExternalIdentityProvider && !this.twoFactorEnabled && this.isActive
    }

    public get canDisableMFA() {
        return !this.providerRequiresMFA && !this.isExternalIdentityProvider && this.twoFactorEnabled && this.isActive
    }

    public get canResetMFA() {
        return !this.isExternalIdentityProvider && this.twoFactorEnabled && this.twoFactorConfigured && this.isActive
    }
    
    public get canClearLockout() {
        return !this.isExternalIdentityProvider && this.lockoutEnd && this.lockoutEnd > new Date() && this.isActive;
    }

    public toString() {
        return `${this.firstName} ${this.lastName}`;
    }
}